### PR TITLE
[Merged by Bors] - More sync updates

### PR DESCRIPTION
**Pull request by divagant-martian**
## Issue Addressed
#1614 and a couple of sync-stalling problems, the most important is a cyclic dependency between the sync manager and the peer manager

### CODE REVIEW

File: beacon_node/network/src/sync/range_sync/chain.rs [hunk 1]
```diff
@@ -340,10 +336,8 @@ impl<T: BeaconChainTypes> SyncingChain<T> {
                         // - Poisoned -> this is an intermediate state that should never be reached
                         // - AwaitingDownload -> A recoverable failed batch should have been
                         //   re-requested.
-                        unreachable!(
-                            "Optimistic batch indicates inconsistent chain state: {:?}",
-                            state
-                        )
+                        debug!(self.log, "Optimistic batch indicates inconsistent chain state"; "state" => ?state);
```
Comment (AgeManning) [inline beacon_node/network/src/sync/range_sync/chain.rs:339]: should we make this a crit if its supposed to be unreachable?

File: beacon_node/network/src/sync/range_sync/chain.rs [hunk 2]
```diff
@@ -610,12 +613,13 @@ impl<T: BeaconChainTypes> SyncingChain<T> {
                         active_batches.remove(&id);
                     }
                 }
-                BatchState::Failed | BatchState::Poisoned | BatchState::AwaitingDownload => {
-                    unreachable!("batch indicates inconsistent chain state while advancing chain")
-                }
+                BatchState::Failed | BatchState::Poisoned | BatchState::AwaitingDownload => debug!(
```
Comment (AgeManning) [inline beacon_node/network/src/sync/range_sync/chain.rs:616]: similarly here, if it was unreachable, i'd expect this to be a crit right? 
Or this this likely to happen?

**Review by AgeManning — COMMENTED** [review]
This looks much cleaner. 

So there was this annoying edge case I wanted to handle which made some of the code not so great. 

Once we complete all finalized chains syncing, its likely the case that the connected peers are all part of the finalized chain. So when we remove them we are in this awkward state where we need to re-status all the peers expecting to start doing some head syncing. 

So I had some logic in there which switches the state to "head syncing" preemptively whilst we wait for the re-status. I wanted to prevent the state from doing finalized -> stalled -> head. When we're not really stalled we're still kind of syncing. 

Wondering if this update still handles this, or do we go into stalled? Wondering what you think we should do with this edge case.

File: beacon_node/network/src/sync/range_sync/chain.rs (hunk 1 above)
Comment (divagant-martian) [reply beacon_node/network/src/sync/range_sync/chain.rs:339]: yep

File: beacon_node/network/src/sync/range_sync/chain.rs (hunk 2 above)
Comment (divagant-martian) [reply beacon_node/network/src/sync/range_sync/chain.rs:616]: no, this is a CRIT too

**Review by AgeManning — APPROVED** [review]
bors r+

**Pull request closed without merging**